virtual_ecosystem/models/animal/animal_model.py

```diff
@@ -203,32 +219,113 @@ def _setup_grid_neighbours(self) -> None:
+
+            for size, cell_id in zip(cohort_sizes, cohort_locations):
                 self.create_new_cohort(
-                    functional_group=functional_group,
-                    mass=functional_group.adult_mass,
+                    functional_group=fg,
+                    mass=fg.adult_mass,
                     age=0.0,
-                    individuals=individuals,
+                    individuals=size,
                     centroid_key=cell_id,
                 )
 
+    def _estimate_total_individuals(self, functional_group: FunctionalGroup) -> int:
+        """Estimates the total number of individuals of a functional group.
+
+        Args:
+            functional_group: The specific functional group having its individuals
+                estimated.
+
+        Returns: The integer number of individuals of the group.
+
+        """
+
+        total_area = self.data.grid.n_cells * self.data.grid.cell_area
+
+        if functional_group.density_individuals_m2 is not None:
+            # User-provided empirical density overrides scaling laws
+            return int(functional_group.density_individuals_m2 * total_area)
+
+        # No empirical density → use selected scaling method
+        if self.density_scaling_method == "damuth":
+            density = damuths_law(
+                functional_group.adult_mass,
+                functional_group.population_density_terms,
+            )
+        elif self.density_scaling_method == "madingley":
+            density = madingley_individuals_density(
+                functional_group.adult_mass,
+                functional_group.population_density_terms,
+            )
+        else:
+            raise ValueError(
+                f"Unsupported density scaling method: {self.density_scaling_method}"
+            )
+
+        return ceil(density * total_area)
+
+    def _distribute_individuals_to_cohorts(self, total_individuals: int) -> list[int]:
+        """Distribute individuals into cohorts respecting minimum size.
+
+        Args:
+            total_individuals: The number of individuals to distribute.
+
+        Returns:
+            A list of cohort sizes.
+        """
+        n_target = self.target_cohorts_per_fg  # ideal number of cohorts
+        min_size = self.minimum_cohort_size  # minimum size of cohorts
+
+        if total_individuals < n_target * min_size:
+            # if I don't have enough individuals to meet my size and number targets
+            # reduce the number of cohorts
+            n_target = max(1, total_individuals // min_size)
+
+        base_size = total_individuals // n_target  # the number of indiv in each cohorts
+        remainder = total_individuals % n_target  # the leftover number of indiv
+
+        # evenly distribute the remained and return the list of cohort sizes
+        return [base_size + 1 if i < remainder else base_size for i in range(n_target)]
+
+    def _assign_cohort_locations(self, n_cohorts: int) -> list[int]:
```

Comment:
Nothing to do with the code, but presumably, at some point the distribution of animals in cells will depend on what ecosystem is the one of that cell - eg. if we have savanna cells and tropical forest cells, put the elephants in the savanna ones, right?

Reply:
Once we get into patchy ecosystems, I will definitely implement something like that. Our initial cases are basically contiguous ecosystems so it works for now.